File: enemy.py
# enemy.py

import os
import pygame
import math
import random
from bullets import Bullet
from settings import TILE_SIZE, resource_path
import map

class Enemy:
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.size = 32
        self.health = 100
        self.speed = 3
        self.angle = 0
        
        # State machine
        self.state = "patrol" # patrol, attack, investigate
        self.target_x = self.x + random.randint(-150, 150)
        self.target_y = self.y + random.randint(-150, 150)
        self.sight_range = 300
        
        self.shoot_timer = 0
        self.shoot_delay = 60 # Frames between shots
        
        image_path = resource_path(os.path.join('assets', 'PNG', 'Soldier 1', 'soldier1_gun.png'))
        self.original_image = pygame.image.load(image_path).convert_alpha()
        self.original_image = pygame.transform.scale(self.original_image, (self.size, self.size))

    def update(self, player, enemies_list):
        # Determine distance to player
        dx = player.x - self.x
        dy = player.y - self.y
        dist_to_player = math.sqrt(dx**2 + dy**2)
        
        if dist_to_player <= self.sight_range and self.has_line_of_sight(player):
            self.state = "attack"
            self.target_x = player.x
            self.target_y = player.y
        elif self.state == "attack": # lost sight
            self.state = "investigate" # go to last known loc

        # Behavior execution
        bullets_to_spawn = None
        if self.state == "attack":
            # face player
            self.angle = -math.degrees(math.atan2(dy, dx))
            
            # handle shooting
            if self.shoot_timer <= 0:
                bullets_to_spawn = Bullet(self.x + self.size//2, self.y + self.size//2, player.x, player.y, 'enemy', 15)
                self.shoot_timer = self.shoot_delay
            else:
                self.shoot_timer -= 1
        else:
            # face target
            tdx = self.target_x - self.x
            tdy = self.target_y - self.y
            self.angle = -math.degrees(math.atan2(tdy, tdx))
            
            # if arrived near target and not attacking
            if math.sqrt(tdx**2 + tdy**2) < 10:
                if self.state == "investigate":
                    self.state = "patrol"
                self.target_x = self.x + random.randint(-150, 150)
                self.target_y = self.y + random.randint(-150, 150)

        # Move towards target
        move_dx = self.target_x - self.x
        move_dy = self.target_y - self.y
        dist = math.sqrt(move_dx**2 + move_dy**2)
        
        if dist > 0:
            move_x = (move_dx / dist) * self.speed
            move_y = (move_dy / dist) * self.speed
            
            new_x = self.x + move_x
            e_rect_x = pygame.Rect(new_x, self.y, self.size, self.size)
            collided_x = False
            for other in enemies_list:
                if other != self:
                    if e_rect_x.colliderect(pygame.Rect(other.x, other.y, other.size, other.size)):
                        collided_x = True
                        break
            if not collided_x and self._can_move(new_x, self.y):
                self.x = new_x
            else:
                self.target_x = self.x + random.randint(-150, 150)
                
            new_y = self.y + move_y
            e_rect_y = pygame.Rect(self.x, new_y, self.size, self.size)
            collided_y = False
            for other in enemies_list:
                if other != self:
                    if e_rect_y.colliderect(pygame.Rect(other.x, other.y, other.size, other.size)):
                        collided_y = True
                        break
            if not collided_y and self._can_move(self.x, new_y):
                self.y = new_y
            else:
                self.target_y = self.y + random.randint(-150, 150)
                
        return bullets_to_spawn
        
    def investigate(self, x, y):
        self.state = "investigate"
        self.target_x = x
        self.target_y = y

    def take_damage(self, amount):
        self.health -= amount

    def _can_move(self, x, y):
        tiles = [
            (y // TILE_SIZE, x // TILE_SIZE),
            (y // TILE_SIZE, (x + self.size) // TILE_SIZE),
            ((y + self.size) // TILE_SIZE, x // TILE_SIZE),
            ((y + self.size) // TILE_SIZE, (x + self.size) // TILE_SIZE)
        ]
        for ty, tx in tiles:
            if not (0 <= ty < len(map.grid) and 0 <= tx < len(map.grid[0])):
                return False
            # Collide with wall (1) or box (4)
            if map.grid[int(ty)][int(tx)] in [1, 4]:
                return False
        return True

    def has_line_of_sight(self, player):
        x1, y1 = self.x + self.size // 2, self.y + self.size // 2
        x2, y2 = player.x + player.size // 2, player.y + player.size // 2

        dist = math.sqrt((x2 - x1)**2 + (y2 - y1)**2)
        if dist == 0: return True
        
        steps = int(dist / (TILE_SIZE / 2))
        if steps == 0: return True
        
        dx = (x2 - x1) / steps
        dy = (y2 - y1) / steps
        
        cx, cy = x1, y1
        for _ in range(steps):
            cx += dx
            cy += dy
            grid_y = int(cy // TILE_SIZE)
            grid_x = int(cx // TILE_SIZE)
            if 0 <= grid_y < len(map.grid) and 0 <= grid_x < len(map.grid[0]):
                if map.grid[grid_y][grid_x] in [1, 4]: # Lines of sight blocked by boxes too
                    return False
        return True

    def draw(self, screen, camera_x, camera_y):
        rotated_image = pygame.transform.rotate(self.original_image, self.angle)
        rect = rotated_image.get_rect(center=(self.x - camera_x + self.size // 2, self.y - camera_y + self.size // 2))
        screen.blit(rotated_image, rect.topleft)
        
        if self.health < 100:
            pygame.draw.rect(screen, (255, 0, 0), (self.x - camera_x, self.y - camera_y - 10, self.size, 5))
            pygame.draw.rect(screen, (0, 255, 0), (self.x - camera_x, self.y - camera_y - 10, self.size * (self.health / 100), 5))

File: Heist-Native/app/src/main/java/org/test/heist/Enemy.kt
package org.test.heist

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import java.util.Random
import kotlin.math.*

class Enemy(val context: Context, var x: Float, var y: Float, private val enemyBitmap: Bitmap?) {
    val size = 80f                      // Slightly smaller hitbox
    var health = 100
    var speed = 3f
    var angle = 0f

    private var state = "patrol"
    private var targetX = x
    private var targetY = y
    private var level = 1
    private val sightRange = 550f
    private val shootRange = 450f       // Only shoot within shorter range to avoid wall shots

    // Stuck detection
    private var stuckTimer = 0
    private var lastX = x
    private var lastY = y

    // Combat
    private var shootTimer = 0
    private var shootDelay = 45

    private val random = Random()
    private val matrix = Matrix()
    private val healthPaint = Paint()

    init {
        targetX = x + random.nextInt(300) - 150
        targetY = y + random.nextInt(300) - 150
    }

    fun setLevel(l: Int) {
        level = l
        speed = 2.5f + (l * 0.15f)
        shootDelay = maxOf(25, 50 - (l * 2))
    }

    fun investigate(ix: Float, iy: Float) {
        if (state != "attack") {
            state = "investigate"
            targetX = ix; targetY = iy
        }
    }

    fun update(playerX: Float, playerY: Float, tileMap: TileMap, bulletBmp: Bitmap?): Bullet? {
        val dx = playerX - x
        val dy = playerY - y
        val distToPlayer = sqrt(dx * dx + dy * dy)

        var spawnedBullet: Bullet? = null

        // ── Sight & Shooting: only fire if there's a clear line-of-sight ──
        val hasLOS = distToPlayer < sightRange && hasLineOfSight(x, y, playerX, playerY, tileMap)

        if (hasLOS) {
            state = "attack"
            targetX = playerX
            targetY = playerY
            angle = Math.toDegrees(atan2(dy.toDouble(), dx.toDouble())).toFloat()

            // Only shoot if ALSO within shoot range (prevents wall-snipe from far away)
            if (distToPlayer < shootRange && shootTimer <= 0) {
                spawnedBullet = Bullet(context, x, y, playerX, playerY, "enemy", bulletBmp)
                shootTimer = shootDelay
            } else if (shootTimer > 0) {
                shootTimer--
            }
        } else {
            if (state == "attack") {
                // Lost sight — move to last known position
                state = "investigate"
            }
            if (shootTimer > 0) shootTimer--
        }

        // ── Movement AI ──
        val tdx = targetX - x
        val tdy = targetY - y
        val distToTarget = sqrt(tdx * tdx + tdy * tdy)

        if (distToTarget > 20f) {
            if (state != "attack") {
                angle = Math.toDegrees(atan2(tdy.toDouble(), tdx.toDouble())).toFloat()
            }

            val nx = x + (tdx / distToTarget) * speed
            val ny = y + (tdy / distToTarget) * speed

            // Multi-point bounding box collision (half-size = 38px)
            val r = size / 2f - 6f
            val moved = when {
                canMoveBox(nx, y, r, tileMap)  -> { x = nx; true }
                canMoveBox(x, ny, r, tileMap)  -> { y = ny; true }
                else -> false
            }

            if (!moved) {
                // Pick a new random target when blocked
                targetX = x + random.nextInt(400) - 200f
                targetY = y + random.nextInt(400) - 200f
                stuckTimer = 0
            }
        } else {
            // Reached target — pick new patrol destination
            if (state == "patrol" || state == "investigate") {
                state = "patrol"
                targetX = x + random.nextInt(400) - 200f
                targetY = y + random.nextInt(400) - 200f
            }
        }

        // ── Stuck detection: if barely moved in 120 frames, get new target ──
        stuckTimer++
        if (stuckTimer >= 120) {
            val movedDist = sqrt((x - lastX).pow(2) + (y - lastY).pow(2))
            if (movedDist < 10f && state != "attack") {
                targetX = x + random.nextInt(500) - 250f
                targetY = y + random.nextInt(500) - 250f
            }
            lastX = x; lastY = y; stuckTimer = 0
        }

        return spawnedBullet
    }

    /**
     * Multi-point AABB collision — checks 4 corners of the enemy's bounding circle.
     * Prevents clipping into walls from any edge.
     */
    private fun canMoveBox(nx: Float, ny: Float, r: Float, tileMap: TileMap): Boolean {
        val ts = tileMap.tileSize
        // Check all 4 corners of the bounding box
        val corners = listOf(
            Pair(nx - r, ny - r), Pair(nx + r, ny - r),
            Pair(nx - r, ny + r), Pair(nx + r, ny + r)
        )
        for ((cx, cy) in corners) {
            val gx = (cx / ts).toInt()
            val gy = (cy / ts).toInt()
            if (gx !in 0 until tileMap.mapWidth || gy !in 0 until tileMap.mapHeight) return false
            val t = tileMap.grid[gy][gx]
            if (t == 1 || t == 4) return false
        }
        return true
    }

    /**
     * DDA ray-cast line-of-sight check.
     * Walks tiles along the line from (x1,y1) to (x2,y2).
     * Returns false if any wall tile (1) is crossed.
     */
    private fun hasLineOfSight(x1: Float, y1: Float, x2: Float, y2: Float, tileMap: TileMap): Boolean {
        val ts = tileMap.tileSize.toFloat()
        val steps = (sqrt((x2 - x1).pow(2) + (y2 - y1).pow(2)) / (ts * 0.5f)).toInt() + 1
        for (i in 0..steps) {
            val t = i.toFloat() / steps
            val sx = x1 + (x2 - x1) * t
            val sy = y1 + (y2 - y1) * t
            val gx = (sx / ts).toInt()
            val gy = (sy / ts).toInt()
            if (gx in 0 until tileMap.mapWidth && gy in 0 until tileMap.mapHeight) {
                if (tileMap.grid[gy][gx] == 1) return false   // Wall blocks LOS
            }
        }
        return true
    }

    fun draw(canvas: Canvas, cameraX: Float, cameraY: Float) {
        val sx = x - cameraX
        val sy = y - cameraY

        if (enemyBitmap != null) {
            matrix.reset()
            // Enemy asset: faces right by default — same convention as player
            matrix.postRotate(angle, enemyBitmap.width / 2f, enemyBitmap.height / 2f)
            matrix.postTranslate(sx - enemyBitmap.width / 2f, sy - enemyBitmap.height / 2f)
            canvas.drawBitmap(enemyBitmap, matrix, null)
        } else {
            // Fallback: draw colored circle soldier
            healthPaint.color = android.graphics.Color.parseColor("#E53935")
            healthPaint.style = Paint.Style.FILL
            canvas.drawCircle(sx, sy, 38f, healthPaint)
            healthPaint.color = android.graphics.Color.parseColor("#B71C1C")
            healthPaint.style = Paint.Style.STROKE; healthPaint.strokeWidth = 4f
            canvas.drawCircle(sx, sy, 38f, healthPaint)
            healthPaint.style = Paint.Style.FILL
            // Direction indicator
            val rad = Math.toRadians(angle.toDouble())
            healthPaint.color = android.graphics.Color.WHITE
            canvas.drawCircle((sx + cos(rad) * 28f).toFloat(), (sy + sin(rad) * 28f).toFloat(), 8f, healthPaint)
        }

        // Health bar (always shown when damaged)
        if (health < 100) {
            healthPaint.color = android.graphics.Color.parseColor("#33FF0000")
            healthPaint.style = Paint.Style.FILL
            canvas.drawRect(sx - 42f, sy - 58f, sx + 42f, sy - 46f, healthPaint)
            healthPaint.color = android.graphics.Color.parseColor("#FF4444")
            canvas.drawRect(sx - 42f, sy - 58f, sx - 42f + (84f * health / 100f), sy - 46f, healthPaint)
        }
    }
}
